--- old/src/hotspot/share/runtime/thread.cpp	2018-09-26 14:03:55.000000000 +0200
+++ new/src/hotspot/share/runtime/thread.cpp	2018-09-26 14:03:55.000000000 +0200
@@ -1549,6 +1549,7 @@
   _pending_failed_speculation = 0;
   _pending_transfer_to_interpreter = false;
   _adjusting_comp_level = false;
+  _in_retryable_allocation = false;
   _jvmci._alternate_call_target = NULL;
   assert(_jvmci._implicit_exception_pc == NULL, "must be");
   if (JVMCICounterSize > 0) {
