--- old/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/CompilerToVM.java	2017-11-03 16:52:59.000000000 +0100
+++ new/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/CompilerToVM.java	2017-11-03 16:52:58.000000000 +0100
@@ -470,7 +470,7 @@
     /**
      * Sets flags on {@code method} indicating that it should never be inlined or compiled by the VM.
      */
-    native void setNotInlineableOrCompileable(HotSpotResolvedJavaMethodImpl method);
+    native void setNotInlinableOrCompilable(HotSpotResolvedJavaMethodImpl method);
 
     /**
      * Invalidates the profiling information for {@code method} and (re)initializes it such that
