--- old/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedJavaMethodImpl.java	2017-11-03 16:53:00.000000000 +0100
+++ new/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedJavaMethodImpl.java	2017-11-03 16:53:00.000000000 +0100
@@ -320,8 +320,8 @@
     /**
      * Sets flags on {@code method} indicating that it should never be inlined or compiled by the VM.
      */
-    public void setNotInlineableOrCompileable() {
-        compilerToVM().setNotInlineableOrCompileable(this);
+    public void setNotInlinableOrCompilable() {
+        compilerToVM().setNotInlinableOrCompilable(this);
     }
 
     /**
