--- old/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedJavaMethod.java	2017-11-03 16:52:59.000000000 +0100
+++ new/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedJavaMethod.java	2017-11-03 16:52:59.000000000 +0100
@@ -59,7 +59,7 @@
     /**
      * Sets flags on {@code method} indicating that it should never be inlined or compiled by the VM.
      */
-    void setNotInlineableOrCompileable();
+    void setNotInlinableOrCompilable();
 
     /**
      * Returns true if this method is one of the special methods that is ignored by security stack
