--- old/src/share/vm/jvmci/jvmciCompilerToVM.hpp	2016-08-23 23:35:38.000000000 +0200
+++ new/src/share/vm/jvmci/jvmciCompilerToVM.hpp	2016-08-23 23:35:38.000000000 +0200
@@ -83,8 +83,10 @@
     }
   };
 
- public:
   static JNINativeMethod methods[];
+
+  static objArrayHandle initialize_intrinsics(TRAPS);
+ public:
   static int methods_count();
 
   static inline Method* asMethod(jobject jvmci_method) {
