--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotSignature.java	2016-05-12 09:40:47.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotSignature.java	2016-05-12 09:40:46.000000000 +0200
@@ -25,7 +25,6 @@
 import java.util.ArrayList;
 import java.util.List;
 
-import jdk.vm.ci.common.JVMCIError;
 import jdk.vm.ci.meta.JavaKind;
 import jdk.vm.ci.meta.JavaType;
 import jdk.vm.ci.meta.ResolvedJavaType;
@@ -105,7 +104,7 @@
             case 'Z':
                 break;
             default:
-                throw new JVMCIError("Invalid character at index %d in signature: %s", cur, signature);
+                throw new InternalError(String.format("Invalid character at index %d in signature: %s", cur, signature));
         }
         return cur;
     }
