--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotConstantPool.java	2016-05-12 09:40:39.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotConstantPool.java	2016-05-12 09:40:39.000000000 +0200
@@ -29,7 +29,6 @@
 
 import java.lang.invoke.MethodHandle;
 
-import jdk.vm.ci.common.JVMCIError;
 import jdk.vm.ci.meta.ConstantPool;
 import jdk.vm.ci.meta.JavaConstant;
 import jdk.vm.ci.meta.JavaField;
@@ -158,7 +157,7 @@
                 if (res != null) {
                     return res;
                 }
-                throw new JVMCIError("Unknown JVM_CONSTANT tag %s", tag);
+                throw new InternalError("Unknown JVM_CONSTANT tag " + tag);
             }
         }
 
@@ -508,7 +507,7 @@
                 Object obj = compilerToVM().resolveConstantInPool(this, cpi);
                 return HotSpotObjectConstantImpl.forObject(obj);
             default:
-                throw new JVMCIError("Unknown constant pool tag %s", tag);
+                throw new InternalError("Unknown constant pool tag " + tag);
         }
     }
 
@@ -655,7 +654,7 @@
                 break;
             }
             default:
-                throw JVMCIError.shouldNotReachHere("Unexpected opcode " + opcode);
+                throw new InternalError("Unexpected opcode " + opcode);
         }
 
         final JVM_CONSTANT tag = getTagAt(index);
