--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotJVMCIRuntimeProvider.java	2016-05-12 09:40:42.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotJVMCIRuntimeProvider.java	2016-05-12 09:40:42.000000000 +0200
@@ -24,13 +24,12 @@
 
 import java.io.OutputStream;
 
-import jdk.vm.ci.common.JVMCIError;
+import jdk.internal.misc.Unsafe;
 import jdk.vm.ci.meta.JVMCIMetaAccessContext;
 import jdk.vm.ci.meta.JavaKind;
 import jdk.vm.ci.meta.JavaType;
 import jdk.vm.ci.meta.ResolvedJavaType;
 import jdk.vm.ci.runtime.JVMCIRuntime;
-import jdk.internal.misc.Unsafe;
 
 //JaCoCo Exclude
 
@@ -98,7 +97,7 @@
             case Object:
                 return Unsafe.ARRAY_OBJECT_BASE_OFFSET;
             default:
-                throw new JVMCIError("%s", kind);
+                throw new InternalError(kind.toString());
         }
     }
 
@@ -128,7 +127,7 @@
             case Object:
                 return Unsafe.ARRAY_OBJECT_INDEX_SCALE;
             default:
-                throw new JVMCIError("%s", kind);
+                throw new InternalError(kind.toString());
         }
     }
 }
