--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedPrimitiveType.java	2016-05-12 09:40:46.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedPrimitiveType.java	2016-05-12 09:40:46.000000000 +0200
@@ -28,7 +28,6 @@
 import java.lang.reflect.Array;
 import java.lang.reflect.Modifier;
 
-import jdk.vm.ci.common.JVMCIError;
 import jdk.vm.ci.meta.Assumptions.AssumptionResult;
 import jdk.vm.ci.meta.JavaConstant;
 import jdk.vm.ci.meta.JavaKind;
@@ -100,7 +99,7 @@
 
     @Override
     public ResolvedJavaType getSingleImplementor() {
-        throw new JVMCIError("Cannot call getSingleImplementor() on a non-interface type: %s", this);
+        throw new InternalError("Cannot call getSingleImplementor() on a non-interface type: " + this);
     }
 
     @Override
@@ -230,7 +229,7 @@
 
     @Override
     public String getSourceFileName() {
-        throw JVMCIError.unimplemented();
+        throw new InternalError();
     }
 
     @Override
