--- old/src/share/vm/runtime/frame.cpp	2013-06-20 15:57:57.996602694 +0200
+++ new/src/share/vm/runtime/frame.cpp	2013-06-20 15:57:57.940602694 +0200
@@ -713,7 +713,7 @@
       Method* m = ((nmethod *)_cb)->method();
       if (m != NULL) {
         m->name_and_sig_as_C_string(buf, buflen);
-        st->print("J  %s", buf);
+        st->print("J  %s @ " PTR_FORMAT " [" PTR_FORMAT "+%d]", buf, _pc, _cb->code_begin(), _pc - _cb->code_begin());
       } else {
         st->print("J  " PTR_FORMAT, pc());
       }
