--- old/src/hotspot/.mx.jvmci/suite.py	2018-01-29 17:58:43.000000000 +0100
+++ new/src/hotspot/.mx.jvmci/suite.py	2018-01-29 17:58:43.000000000 +0100
@@ -208,6 +208,17 @@
       "workingSets" : "JVMCI,HotSpot,SPARC",
     },
 
+    "jdk.vm.ci.hotspot.jfr" : {
+      "subDir" : "../jdk.internal.vm.ci/share/classes",
+      "sourceDirs" : ["src"],
+      "dependencies" : [
+        "jdk.vm.ci.hotspot",
+      ],
+      "checkstyle" : "jdk.vm.ci.services",
+      "javaCompliance" : "9",
+      "workingSets" : "JVMCI,HotSpot",
+    },
+
     "hotspot" : {
       "native" : True,
       "class" : "HotSpotProject",
