--- old/src/share/vm/jvmci/jvmciCodeInstaller.cpp	2016-07-14 23:59:39.000000000 +0200
+++ new/src/share/vm/jvmci/jvmciCodeInstaller.cpp	2016-07-14 23:59:39.000000000 +0200
@@ -782,7 +782,7 @@
     }
     last_pc_offset = pc_offset;
 
-    if (CodeInstallSafepointChecks && SafepointSynchronize::do_call_back()) {
+    if (SafepointSynchronize::do_call_back()) {
       // this is a hacky way to force a safepoint check but nothing else was jumping out at me.
       ThreadToNativeFromVM ttnfv(JavaThread::current());
     }
