--- old/src/hotspot/share/oops/annotations.hpp	2018-11-20 10:04:09.000000000 +0100
+++ new/src/hotspot/share/oops/annotations.hpp	2018-11-20 10:04:09.000000000 +0100
@@ -42,6 +42,8 @@
 // a type_annotation instance.
 
 class Annotations: public MetaspaceObj {
+ friend class JVMCIVMStructs;
+
   // If you add a new field that points to any metaspace object, you
   // must add this field to Annotations::metaspace_pointers_do().
 
