--- old/src/share/vm/runtime/globals.hpp	2016-08-11 15:25:51.000000000 +0200
+++ new/src/share/vm/runtime/globals.hpp	2016-08-11 15:25:51.000000000 +0200
@@ -3027,9 +3027,6 @@
           "Number of receiver types to record in call/cast profile")        \
           range(0, 8)                                                       \
                                                                             \
-  experimental(intx, MethodProfileWidth, 0,                                 \
-          "Number of methods to record in call profile")                    \
-                                                                            \
   develop(intx, BciProfileWidth,      2,                                    \
           "Number of return bci's to record in ret profile")                \
                                                                             \
