--- old/test/hotspot/jtreg/compiler/jvmci/compilerToVM/AllocateCompileIdTest.java	2018-09-18 09:55:32.000000000 +0200
+++ new/test/hotspot/jtreg/compiler/jvmci/compilerToVM/AllocateCompileIdTest.java	2018-09-18 09:55:32.000000000 +0200
@@ -33,7 +33,8 @@
  *          jdk.internal.vm.ci/jdk.vm.ci.hotspot
  *          jdk.internal.vm.ci/jdk.vm.ci.code
  *
- * @build jdk.internal.vm.ci/jdk.vm.ci.hotspot.CompilerToVMHelper sun.hotspot.WhiteBox
+ * @build jdk.internal.vm.ci/jdk.vm.ci.hotspot.CompilerToVMHelper
+ *        sun.hotspot.WhiteBox sun.hotspot.parser.DiagnosticCommand
  * @run driver ClassFileInstaller sun.hotspot.WhiteBox
  *                                sun.hotspot.parser.DiagnosticCommand
  *                                sun.hotspot.WhiteBox$WhiteBoxPermission
