--- old/test/compiler/jvmci/events/JvmciShutdownEventListener.java	2016-05-10 13:48:53.000000000 +0200
+++ new/test/compiler/jvmci/events/JvmciShutdownEventListener.java	2016-05-10 13:48:53.000000000 +0200
@@ -24,9 +24,9 @@
 package compiler.jvmci.events;
 
 import jdk.vm.ci.hotspot.HotSpotJVMCIRuntime;
-import jdk.vm.ci.hotspot.HotSpotVMEventListener;
+import jdk.vm.ci.hotspot.services.HotSpotVMEventListener;
 
-public class JvmciShutdownEventListener implements HotSpotVMEventListener {
+public class JvmciShutdownEventListener extends HotSpotVMEventListener {
     public static final String MESSAGE = "Shutdown notified";
     public static final String GOT_INTERNAL_ERROR = "Got internal error";
 
