--- old/test/hotspot/jtreg/compiler/jvmci/compilerToVM/DoNotInlineOrCompileTest.java	2017-11-03 16:53:02.000000000 +0100
+++ new/test/hotspot/jtreg/compiler/jvmci/compilerToVM/DoNotInlineOrCompileTest.java	2017-11-03 16:53:02.000000000 +0100
@@ -71,10 +71,10 @@
         boolean hasNeverInlineDirective = CompilerToVMHelper.hasNeverInlineDirective(method);
         Asserts.assertFalse(hasNeverInlineDirective, "Unexpected initial " +
                 "value of property 'hasNeverInlineDirective'");
-        CompilerToVMHelper.setNotInlineableOrCompileable(method);
+        CompilerToVMHelper.setNotInlinableOrCompilable(method);
         hasNeverInlineDirective = CompilerToVMHelper.hasNeverInlineDirective(method);
         Asserts.assertTrue(hasNeverInlineDirective, aMethod
-                + " : hasNeverInlineDirective is false even after setNotInlineableOrCompileable'");
+                + " : hasNeverInlineDirective is false even after setNotInlinableOrCompilable'");
     }
 
     private static List<Executable> createTestCases() {
