--- old/src/hotspot/share/runtime/thread.cpp	2018-10-05 16:54:24.000000000 +0200
+++ new/src/hotspot/share/runtime/thread.cpp	2018-10-05 16:54:23.000000000 +0200
@@ -1547,6 +1547,7 @@
   _pending_failed_speculation = 0;
   _pending_transfer_to_interpreter = false;
   _adjusting_comp_level = false;
+  _in_retryable_allocation = false;
   _jvmci._alternate_call_target = NULL;
   assert(_jvmci._implicit_exception_pc == NULL, "must be");
   if (JVMCICounterSize > 0) {
