--- old/src/share/vm/jvmci/jvmciRuntime.hpp	2016-07-15 00:39:33.000000000 +0200
+++ new/src/share/vm/jvmci/jvmciRuntime.hpp	2016-07-15 00:39:32.000000000 +0200
@@ -87,6 +87,11 @@
   static Handle callStatic(const char* className, const char* methodName, const char* returnType, JavaCallArguments* args, TRAPS);
 
   /**
+   * Determines if the VM is sufficiently booted to initialize JVMCI.
+   */
+  static bool can_initialize_JVMCI();
+
+  /**
    * Trigger initialization of HotSpotJVMCIRuntime through JVMCI.getRuntime()
    */
   static void initialize_JVMCI(TRAPS);
