--- old/src/share/vm/jvmci/jvmci_globals.cpp	2016-10-05 16:56:34.000000000 +0200
+++ new/src/share/vm/jvmci/jvmci_globals.cpp	2016-10-05 16:56:34.000000000 +0200
@@ -85,6 +85,7 @@
   CHECK_NOT_SET(JVMCIUseFastLocking,          EnableJVMCI)
   CHECK_NOT_SET(JVMCINMethodSizeLimit,        EnableJVMCI)
   CHECK_NOT_SET(MethodProfileWidth,           EnableJVMCI)
+  CHECK_NOT_SET(JVMCIPrintSystemProperties,   EnableJVMCI)
   CHECK_NOT_SET(TraceUncollectedSpeculations, EnableJVMCI)
 
 #ifndef PRODUCT
