--- old/make/common/Modules.gmk	2017-04-28 08:02:57.000000000 +0200
+++ new/make/common/Modules.gmk	2017-04-28 08:02:56.000000000 +0200
@@ -79,11 +79,14 @@
     java.xml.bind \
     java.xml.ws \
     java.xml.ws.annotation \
+    jdk.internal.vm.compiler \
     #
 
 # Modules explicitly declared as not being upgradeable even though they require
 # an upgradeable module.
-NON_UPGRADEABLE_MODULES +=
+NON_UPGRADEABLE_MODULES += \
+    jdk.aot \
+    #
 
 AGGREGATOR_MODULES += \
     java.se \
@@ -109,7 +112,6 @@
     jdk.crypto.ec \
     jdk.dynalink \
     jdk.incubator.httpclient \
-    jdk.internal.vm.compiler \
     jdk.jsobject \
     jdk.localedata \
     jdk.naming.dns \
