--- old/src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp	2018-06-26 20:47:37.000000000 +0200
+++ new/src/hotspot/share/jvmci/jvmciCompilerToVMInit.cpp	2018-06-26 20:47:36.000000000 +0200
@@ -234,10 +234,15 @@
   do_bool_flag(UseCRC32Intrinsics)                                         \
   do_bool_flag(UseCompressedClassPointers)                                 \
   do_bool_flag(UseCompressedOops)                                          \
-  do_bool_flag(UseConcMarkSweepGC)                                         \
   X86_ONLY(do_bool_flag(UseCountLeadingZerosInstruction))                  \
   X86_ONLY(do_bool_flag(UseCountTrailingZerosInstruction))                 \
+  do_bool_flag(UseConcMarkSweepGC)                                         \
   do_bool_flag(UseG1GC)                                                    \
+  do_bool_flag(UseParallelGC)                                              \
+  do_bool_flag(UseParallelOldGC)                                           \
+  do_bool_flag(UseSerialGC)                                                \
+  do_bool_flag(UseZGC)                                                     \
+  do_bool_flag(UseEpsilonGC)                                               \
   COMPILER2_PRESENT(do_bool_flag(UseMontgomeryMultiplyIntrinsic))          \
   COMPILER2_PRESENT(do_bool_flag(UseMontgomerySquareIntrinsic))            \
   COMPILER2_PRESENT(do_bool_flag(UseMulAddIntrinsic))                      \
