--- old/src/jdk.internal.vm.compiler.management/share/classes/module-info.java	2018-04-18 11:08:35.000000000 +0200
+++ new/src/jdk.internal.vm.compiler.management/share/classes/module-info.java	2018-04-18 11:08:34.000000000 +0200
@@ -34,8 +34,5 @@
     requires jdk.management;
     requires jdk.internal.vm.ci;
     requires jdk.internal.vm.compiler;
-
-    provides sun.management.spi.PlatformMBeanProvider with
-        org.graalvm.compiler.hotspot.jmx.GraalMBeans;
 }
 
