--- old/test/compiler/jvmci/jdk.vm.ci.runtime.test/src/jdk/vm/ci/runtime/test/NameAndSignature.java	2016-02-04 11:05:43.000000000 +0100
+++ new/test/compiler/jvmci/jdk.vm.ci.runtime.test/src/jdk/vm/ci/runtime/test/NameAndSignature.java	2016-02-04 11:05:43.000000000 +0100
@@ -39,7 +39,7 @@
     final Class<?> returnType;
     final Class<?>[] parameterTypes;
 
-    public NameAndSignature(Method m) {
+    NameAndSignature(Method m) {
         this.name = m.getName();
         this.returnType = m.getReturnType();
         this.parameterTypes = m.getParameterTypes();
