--- old/src/os/linux/vm/gpu_linux.cpp	2014-02-03 11:38:08.000000000 +0100
+++ new/src/os/linux/vm/gpu_linux.cpp	2014-02-03 11:38:08.000000000 +0100
@@ -40,7 +40,7 @@
   bool hsail = false;
   bool ptx = false;
 
-  if (UseHSAILSimulator && gpu::Hsail::register_natives(env)) {
+  if (gpu::Hsail::register_natives(env)) {
     hsail = true;
   }
 
--- old/src/os/windows/vm/gpu_windows.cpp	2014-02-03 11:38:09.000000000 +0100
+++ new/src/os/windows/vm/gpu_windows.cpp	2014-02-03 11:38:09.000000000 +0100
@@ -28,7 +28,7 @@
 
 jobject gpu::probe_gpus(JNIEnv* env) {
   // TODO: add detection of PTX/NVidia
-  if (UseHSAILSimulator && gpu::Hsail::register_natives(env)) {
+  if (gpu::Hsail::register_natives(env)) {
     return env->NewStringUTF("HSAIL");
   }
   return env->NewStringUTF("");
--- old/src/share/vm/runtime/globals.hpp	2014-02-03 11:38:10.000000000 +0100
+++ new/src/share/vm/runtime/globals.hpp	2014-02-03 11:38:09.000000000 +0100
@@ -3840,8 +3840,8 @@
   product(bool , AllowNonVirtualCalls, false,                               \
           "Obey the ACC_SUPER flag and allow invokenonvirtual calls")       \
                                                                             \
-  product(bool, UseHSAILSimulator, false,                                   \
-          "Run code on HSAIL Simulator")                                    \
+  product(bool, UseHSAILSimulator, true,                                    \
+          "Use HSAIL simulator if available (fall back to HSAIL hardware)") \
                                                                             \
   diagnostic(ccstr, SharedArchiveFile, NULL,                                \
           "Override the default location of the CDS archive file")          \
