--- old/src/share/vm/classfile/systemDictionary.hpp	2016-10-07 17:17:07.000000000 +0200
+++ new/src/share/vm/classfile/systemDictionary.hpp	2016-10-07 17:17:07.000000000 +0200
@@ -226,7 +226,7 @@
     WKID_LIMIT,
 
 #if INCLUDE_JVMCI
-    FIRST_JVMCI_WKID = WK_KLASS_ENUM_NAME(HotSpotCompiledCode_klass),
+    FIRST_JVMCI_WKID = WK_KLASS_ENUM_NAME(JVMCI_klass),
     LAST_JVMCI_WKID  = WK_KLASS_ENUM_NAME(Value_klass),
 #endif
 
