--- old/test/compiler/jvmci/common/patches/jdk.internal.vm.ci/jdk/vm/ci/hotspot/CompilerToVMHelper.java	2017-08-23 23:10:10.000000000 +0200
+++ new/test/compiler/jvmci/common/patches/jdk.internal.vm.ci/jdk/vm/ci/hotspot/CompilerToVMHelper.java	2017-08-23 23:10:10.000000000 +0200
@@ -79,10 +79,19 @@
     }
 
     public static HotSpotResolvedObjectType lookupType(String name,
-            Class<?> accessingClass, boolean resolve) {
+            Class<?> accessingClass, boolean resolve) throws ClassNotFoundException {
         return CTVM.lookupType(name, accessingClass, resolve);
     }
 
+    public static HotSpotResolvedObjectType lookupTypeHelper(String name,
+            Class<?> accessingClass, boolean resolve) {
+        try {
+            return lookupType(name, accessingClass, resolve);
+        } catch (ClassNotFoundException e) {
+            throw (NoClassDefFoundError) new NoClassDefFoundError().initCause(e);
+        }
+    }
+
     public static Object resolveConstantInPool(ConstantPool constantPool, int cpi) {
         return CTVM.resolveConstantInPool((HotSpotConstantPool) constantPool, cpi);
     }
