--- old/src/jdk.internal.vm.ci/share/classes/module-info.java	2017-04-20 21:50:05.000000000 +0200
+++ new/src/jdk.internal.vm.ci/share/classes/module-info.java	2017-04-20 21:50:05.000000000 +0200
@@ -24,7 +24,7 @@
  */
 
 module jdk.internal.vm.ci {
-    exports jdk.vm.ci.services to jdk.internal.vm.compiler;
+    exports jdk.vm.ci.services;
 
     uses jdk.vm.ci.services.JVMCIServiceLocator;
     uses jdk.vm.ci.hotspot.HotSpotJVMCIBackendFactory;
