--- old/test/hotspot/jtreg/compiler/jvmci/common/patches/jdk.internal.vm.ci/jdk/vm/ci/hotspot/CompilerToVMHelper.java	2017-11-03 16:53:01.000000000 +0100
+++ new/test/hotspot/jtreg/compiler/jvmci/common/patches/jdk.internal.vm.ci/jdk/vm/ci/hotspot/CompilerToVMHelper.java	2017-11-03 16:53:01.000000000 +0100
@@ -224,8 +224,8 @@
         return CTVM.getLocalVariableTableStart((HotSpotResolvedJavaMethodImpl)method);
     }
 
-    public static void setNotInlineableOrCompileable(HotSpotResolvedJavaMethod method) {
-        CTVM.setNotInlineableOrCompileable((HotSpotResolvedJavaMethodImpl)method);
+    public static void setNotInlinableOrCompilable(HotSpotResolvedJavaMethod method) {
+        CTVM.setNotInlinableOrCompilable((HotSpotResolvedJavaMethodImpl)method);
     }
 
     public static void reprofile(HotSpotResolvedJavaMethod method) {
