--- old/src/share/vm/jvmci/jvmci_globals.hpp	2016-10-05 21:59:48.000000000 +0200
+++ new/src/share/vm/jvmci/jvmci_globals.hpp	2016-10-05 21:59:48.000000000 +0200
@@ -49,6 +49,9 @@
   experimental(bool, UseJVMCICompiler, false,                               \
           "Use JVMCI as the default compiler")                              \
                                                                             \
+  experimental(bool, JVMCIPrintProperties, false,                           \
+          "Prints properties used by the JVMCI compiler")                   \
+                                                                            \
   experimental(bool, BootstrapJVMCI, false,                                 \
           "Bootstrap JVMCI before running Java main method")                \
                                                                             \
