--- old/src/share/vm/gc_interface/collectedHeap.hpp	2014-04-11 16:41:50.728234457 +0200
+++ new/src/share/vm/gc_interface/collectedHeap.hpp	2014-04-11 16:41:50.668234459 +0200
@@ -84,6 +84,7 @@
 class CollectedHeap : public CHeapObj<mtInternal> {
   friend class VMStructs;
   friend class IsGCActiveMark; // Block structured external access to _is_gc_active
+  friend class Hsail;  // access to allocate_new_tlab
 
 #ifdef ASSERT
   static int       _fire_out_of_memory_count;
