--- old/test/jdk/jdk/modules/etc/UpgradeableModules.java	2018-06-14 18:12:49.000000000 +0200
+++ new/test/jdk/jdk/modules/etc/UpgradeableModules.java	2018-06-14 18:12:49.000000000 +0200
@@ -45,6 +45,7 @@
     private static final List<String> UPGRADEABLE_MODULES =
         List.of("java.compiler",
                 "java.jnlp",
+                "jdk.aot",
                 "jdk.internal.vm.compiler",
                 "jdk.internal.vm.compiler.management",
                 "jdk.deploy",
