--- old/src/hotspot/share/prims/jni.cpp	2017-11-03 18:40:53.000000000 +0100
+++ new/src/hotspot/share/prims/jni.cpp	2017-11-03 18:40:53.000000000 +0100
@@ -3949,7 +3949,7 @@
         // JVMCI is initialized on a CompilerThread
         if (BootstrapJVMCI) {
           JavaThread* THREAD = thread;
-          JVMCICompiler* compiler = JVMCICompiler::instance(CATCH);
+          JVMCICompiler* compiler = JVMCICompiler::instance(true, CATCH);
           compiler->bootstrap(THREAD);
           if (HAS_PENDING_EXCEPTION) {
             HandleMark hm;
