--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.meta/src/jdk/vm/ci/meta/VMConstant.java	2016-05-18 23:02:16.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.meta/src/jdk/vm/ci/meta/VMConstant.java	2016-05-18 23:02:16.000000000 +0200
@@ -22,5 +22,8 @@
  */
 package jdk.vm.ci.meta;
 
+/**
+ * Represents a constant that needs to be patched at runtime by the VM.
+ */
 public interface VMConstant extends Constant {
 }
