--- old/test/compiler/jvmci/compilerToVM/ReadConfigurationTest.java	2017-04-28 11:49:28.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/ReadConfigurationTest.java	2017-04-28 11:49:28.000000000 +0200
@@ -32,6 +32,7 @@
  * @build jdk.internal.vm.ci/jdk.vm.ci.hotspot.CompilerToVMHelper
  * @build compiler.jvmci.compilerToVM.ReadConfigurationTest
  * @run main/othervm -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI
+ *                   -Djvmci.Compiler=null
  *                   compiler.jvmci.compilerToVM.ReadConfigurationTest
  */
 
