--- old/src/share/vm/jvmci/jvmci_globals.cpp	2016-08-11 15:25:49.000000000 +0200
+++ new/src/share/vm/jvmci/jvmci_globals.cpp	2016-08-11 15:25:49.000000000 +0200
@@ -84,6 +84,7 @@
   CHECK_NOT_SET(JVMCICountersExcludeCompiler, EnableJVMCI)
   CHECK_NOT_SET(JVMCIUseFastLocking,          EnableJVMCI)
   CHECK_NOT_SET(JVMCINMethodSizeLimit,        EnableJVMCI)
+  CHECK_NOT_SET(MethodProfileWidth,           EnableJVMCI)
   CHECK_NOT_SET(TraceUncollectedSpeculations, EnableJVMCI)
 
 #ifndef PRODUCT
