--- old/src/java.base/share/lib/security/default.policy	2017-02-01 20:58:40.000000000 +0100
+++ new/src/java.base/share/lib/security/default.policy	2017-02-01 20:58:40.000000000 +0100
@@ -213,3 +213,7 @@
     permission java.lang.RuntimePermission "accessClassInPackage.com.sun.java.swing.plaf.*";
     permission java.lang.RuntimePermission "accessClassInPackage.com.apple.*";
 };
+
+grant codeBase "jrt:/jdk.vm.compiler" {
+    permission java.security.AllPermission;
+};
