--- old/src/share/vm/jvmci/jvmciCompiler.cpp	2016-05-10 13:48:49.000000000 +0200
+++ new/src/share/vm/jvmci/jvmciCompiler.cpp	2016-05-10 13:48:49.000000000 +0200
@@ -189,6 +189,15 @@
   }
 }
 
+CompLevel JVMCIRuntime::adjust_comp_level(methodHandle method, bool is_osr, CompLevel level, JavaThread* thread) {
+  if (!thread->adjusting_comp_level()) {
+    thread->set_adjusting_comp_level(true);
+    level = adjust_comp_level_inner(method, is_osr, level, thread);
+    thread->set_adjusting_comp_level(false);
+  }
+  return level;
+}
+
 /**
  * Aborts the VM due to an unexpected exception.
  */
