--- old/src/hotspot/.mx.jvmci/suite.py	2018-11-06 17:30:56.000000000 +0100
+++ new/src/hotspot/.mx.jvmci/suite.py	2018-11-06 17:30:56.000000000 +0100
@@ -43,6 +43,9 @@
     "jdk.vm.ci.services" : {
       "subDir" : "../jdk.internal.vm.ci/share/classes",
       "sourceDirs" : ["src"],
+      "imports" : [
+        "jdk.internal.misc"
+      ],
       "javaCompliance" : "9+",
       "checkstyleVersion" : "8.8",
       "workingSets" : "API,JVMCI",
