--- old/make/common/Modules.gmk	2018-04-18 11:08:30.000000000 +0200
+++ new/make/common/Modules.gmk	2018-04-18 11:08:29.000000000 +0200
@@ -76,6 +76,7 @@
 UPGRADEABLE_MODULES += \
     java.compiler \
     jdk.internal.vm.compiler \
+    jdk.internal.vm.compiler.management \
     #
 
 
@@ -103,7 +104,6 @@
     jdk.crypto.ec \
     jdk.dynalink \
     jdk.httpserver \
-    jdk.internal.vm.compiler.management \
     jdk.jsobject \
     jdk.localedata \
     jdk.naming.dns \
@@ -189,6 +189,7 @@
     jdk.hotspot.agent \
     jdk.internal.vm.ci \
     jdk.internal.vm.compiler \
+    jdk.internal.vm.compiler.management \
     #
 
 ################################################################################
