--- old/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.meta/src/jdk/vm/ci/meta/Assumptions.java	2018-01-29 17:58:46.000000000 +0100
+++ new/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.meta/src/jdk/vm/ci/meta/Assumptions.java	2018-01-29 17:58:45.000000000 +0100
@@ -100,7 +100,7 @@
     }
 
     /**
-     * An assumption that a given class has no subclasses implementing {@link Object#finalize()}).
+     * An assumption that a given class has no subclasses implementing {@code Object.finalize()}).
      */
     public static final class NoFinalizableSubclass extends Assumption {
 
