--- old/src/hotspot/.mx.jvmci/.pydevproject	2017-10-27 22:55:47.000000000 +0200
+++ new/src/hotspot/.mx.jvmci/.pydevproject	2017-10-27 22:55:47.000000000 +0200
@@ -2,11 +2,4 @@
 <?eclipse-pydev version="1.0"?><pydev_project>
 <pydev_property name="org.python.pydev.PYTHON_PROJECT_INTERPRETER">Default</pydev_property>
 <pydev_property name="org.python.pydev.PYTHON_PROJECT_VERSION">python 2.7</pydev_property>
-<pydev_pathproperty name="org.python.pydev.PROJECT_SOURCE_PATH">
-<path>/mx.jvmci</path>
-</pydev_pathproperty>
-<pydev_pathproperty name="org.python.pydev.PROJECT_SOURCE_PATH">
-<path>/mx</path>
-</pydev_pathproperty>
-
 </pydev_project>
