--- old/test/compiler/jvmci/jdk.vm.ci.hotspot.test/src/jdk/vm/ci/hotspot/test/AsJavaTypeDataProvider.java	2016-05-10 14:31:32.000000000 +0200
+++ new/test/compiler/jvmci/jdk.vm.ci.hotspot.test/src/jdk/vm/ci/hotspot/test/AsJavaTypeDataProvider.java	2016-05-10 14:31:32.000000000 +0200
@@ -34,26 +34,26 @@
     @DataProvider(name = "asJavaTypeDataProvider")
     public static Object[][] asJavaTypeDataProvider() {
         return new Object[][]{
-                {CONSTANT_REFLECTION_PROVIDER.forObject(DummyClass.class),
-                        "jdk.vm.ci.hotspot.test.DummyClass"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(boolean.class), "boolean"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(byte.class), "byte"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(short.class), "short"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(char.class), "char"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(int.class), "int"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(long.class), "long"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(float.class), "float"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(double.class), "double"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(Object.class), "java.lang.Object"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(boolean[].class), "boolean[]"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(boolean[][].class), "boolean[][]"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(Object[].class), "java.lang.Object[]"},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(Object[][].class), "java.lang.Object[][]"},
-                {JavaConstant.forBoolean(DUMMY_CLASS_INSTANCE.booleanField), null},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(DUMMY_CLASS_INSTANCE.objectField), null},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(DUMMY_CLASS_INSTANCE), null},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(DUMMY_CLASS_INSTANCE.booleanArrayWithValues), null},
-                {CONSTANT_REFLECTION_PROVIDER.forObject(DUMMY_CLASS_INSTANCE.booleanArrayArrayWithValues), null},
-                {JavaConstant.NULL_POINTER, null}, {null, null}};
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(DummyClass.class),
+                                        "jdk.vm.ci.hotspot.test.DummyClass"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(boolean.class), "boolean"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(byte.class), "byte"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(short.class), "short"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(char.class), "char"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(int.class), "int"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(long.class), "long"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(float.class), "float"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(double.class), "double"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(Object.class), "java.lang.Object"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(boolean[].class), "boolean[]"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(boolean[][].class), "boolean[][]"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(Object[].class), "java.lang.Object[]"},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(Object[][].class), "java.lang.Object[][]"},
+                        {JavaConstant.forBoolean(DUMMY_CLASS_INSTANCE.booleanField), null},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(DUMMY_CLASS_INSTANCE.objectField), null},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(DUMMY_CLASS_INSTANCE), null},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(DUMMY_CLASS_INSTANCE.booleanArrayWithValues), null},
+                        {CONSTANT_REFLECTION_PROVIDER.forObject(DUMMY_CLASS_INSTANCE.booleanArrayArrayWithValues), null},
+                        {JavaConstant.NULL_POINTER, null}, {null, null}};
     }
 }
