--- old/test/compiler/jvmci/jdk.vm.ci.hotspot.test/src/jdk/vm/ci/hotspot/test/ForStringDataProvider.java	2016-05-10 14:31:34.000000000 +0200
+++ new/test/compiler/jvmci/jdk.vm.ci.hotspot.test/src/jdk/vm/ci/hotspot/test/ForStringDataProvider.java	2016-05-10 14:31:34.000000000 +0200
@@ -31,8 +31,8 @@
     @DataProvider(name = "forStringDataProvider")
     public static Object[][] forStringDataProvider() {
         return new Object[][]{
-                {DUMMY_CLASS_INSTANCE.stringField, "Object[String:\"" + DUMMY_CLASS_INSTANCE.stringField + "\"]"},
-                {DUMMY_CLASS_INSTANCE.stringEmptyField, "Object[String:\"\"]"},
-                {null, "Object[null]"}};
+                        {DUMMY_CLASS_INSTANCE.stringField, "Object[String:\"" + DUMMY_CLASS_INSTANCE.stringField + "\"]"},
+                        {DUMMY_CLASS_INSTANCE.stringEmptyField, "Object[String:\"\"]"},
+                        {null, "Object[null]"}};
     }
 }
