--- old/.mx.jvmci/mx_jvmci.py	2016-06-21 23:03:39.000000000 +0200
+++ new/.mx.jvmci/mx_jvmci.py	2016-06-21 23:03:38.000000000 +0200
@@ -186,7 +186,8 @@
             # Might be building with JDK8 which has cacerts under jre/
             srcCerts = join(mx.get_jdk(tag='default').home, 'jre', 'lib', 'security', 'cacerts')
         dstCerts = join(jdkImageDir, 'lib', 'security', 'cacerts')
-        shutil.copyfile(srcCerts, dstCerts)
+        if srcCerts != dstCerts:
+            shutil.copyfile(srcCerts, dstCerts)
 
         _create_jdk_bundle(jdkBuildDir, _vm.debugLevel, jdkImageDir)
 
