--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/Stable.java	2016-05-10 14:31:27.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/Stable.java	2016-05-10 14:31:27.000000000 +0200
@@ -29,8 +29,9 @@
 import java.lang.annotation.Target;
 
 /**
- * This annotation functions as an alias for the jdk.internal.vm.annotation.Stable annotation within JVMCI
- * code. It is specially recognized during class file parsing in the same way as that annotation.
+ * This annotation functions as an alias for the jdk.internal.vm.annotation.Stable annotation within
+ * JVMCI code. It is specially recognized during class file parsing in the same way as that
+ * annotation.
  */
 @Target(ElementType.FIELD)
 @Retention(RetentionPolicy.RUNTIME)
