--- old/src/jdk.internal.vm.ci/share/classes/module-info.java	2018-01-29 17:58:47.000000000 +0100
+++ new/src/jdk.internal.vm.ci/share/classes/module-info.java	2018-01-29 17:58:47.000000000 +0100
@@ -24,6 +24,7 @@
  */
 
 module jdk.internal.vm.ci {
+    requires jdk.jfr;
     exports jdk.vm.ci.services to jdk.internal.vm.compiler;
     exports jdk.vm.ci.runtime to
         jdk.internal.vm.compiler,
@@ -36,4 +37,7 @@
         jdk.vm.ci.hotspot.aarch64.AArch64HotSpotJVMCIBackendFactory,
         jdk.vm.ci.hotspot.amd64.AMD64HotSpotJVMCIBackendFactory,
         jdk.vm.ci.hotspot.sparc.SPARCHotSpotJVMCIBackendFactory;
+
+    provides jdk.vm.ci.services.JVMCIServiceLocator with
+        jdk.vm.ci.hotspot.jfr.events.JFREventProvider.Locator;
 }
