--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot.aarch64/src/jdk/vm/ci/hotspot/aarch64/AArch64HotSpotRegisterConfig.java	2016-05-12 09:40:37.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot.aarch64/src/jdk/vm/ci/hotspot/aarch64/AArch64HotSpotRegisterConfig.java	2016-05-12 09:40:36.000000000 +0200
@@ -64,7 +64,6 @@
 import jdk.vm.ci.code.RegisterConfig;
 import jdk.vm.ci.code.StackSlot;
 import jdk.vm.ci.code.TargetDescription;
-import jdk.vm.ci.common.JVMCIError;
 import jdk.vm.ci.hotspot.HotSpotCallingConventionType;
 import jdk.vm.ci.hotspot.HotSpotVMConfig;
 import jdk.vm.ci.meta.AllocatableValue;
@@ -226,7 +225,7 @@
             case Double:
                 return simdParameterRegisters;
             default:
-                throw JVMCIError.shouldNotReachHere();
+                throw new InternalError("should not reach here");
         }
     }
 
@@ -261,7 +260,7 @@
                     }
                     break;
                 default:
-                    throw JVMCIError.shouldNotReachHere();
+                    throw new InternalError("should not reach here");
             }
 
             if (locations[i] == null) {
