--- old/test/compiler/jvmci/compilerToVM/GetBytecodeTest.java	2017-04-21 12:14:58.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/GetBytecodeTest.java	2017-04-21 12:14:58.000000000 +0200
@@ -34,6 +34,7 @@
  *          jdk.internal.vm.ci/jdk.vm.ci.code
  * @build jdk.internal.vm.ci/jdk.vm.ci.hotspot.CompilerToVMHelper
  * @run main/othervm -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI
+ *                   -Djvmci.Compiler=null
  *                   compiler.jvmci.compilerToVM.GetBytecodeTest
  */
 
