--- old/test/compiler/jvmci/compilerToVM/HasFinalizableSubclassTest.java	2017-04-21 12:15:07.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/HasFinalizableSubclassTest.java	2017-04-21 12:15:07.000000000 +0200
@@ -31,7 +31,8 @@
  * @modules jdk.internal.vm.ci/jdk.vm.ci.hotspot
  * @build jdk.internal.vm.ci/jdk.vm.ci.hotspot.CompilerToVMHelper
  * @run main/othervm -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI
- *                  compiler.jvmci.compilerToVM.HasFinalizableSubclassTest
+ *                   -Djvmci.Compiler=null
+ *                   compiler.jvmci.compilerToVM.HasFinalizableSubclassTest
  */
 
 package compiler.jvmci.compilerToVM;
