--- old/test/compiler/jvmci/compilerToVM/HasNeverInlineDirectiveTest.java	2017-04-21 12:15:08.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/HasNeverInlineDirectiveTest.java	2017-04-21 12:15:08.000000000 +0200
@@ -39,6 +39,7 @@
  * @run main/othervm -Xbootclasspath/a:.
  *                   -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI
  *                   -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI
+ *                   -Djvmci.Compiler=null
  *                   compiler.jvmci.compilerToVM.HasNeverInlineDirectiveTest
  */
 
