--- old/test/compiler/jvmci/compilerToVM/IsMatureVsReprofileTest.java	2017-04-21 12:15:10.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/IsMatureVsReprofileTest.java	2017-04-21 12:15:10.000000000 +0200
@@ -38,6 +38,7 @@
  *                                sun.hotspot.WhiteBox$WhiteBoxPermission
  * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions
  *     -XX:+WhiteBoxAPI -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI -Xbatch
+ *     -Djvmci.Compiler=null
  *     compiler.jvmci.compilerToVM.IsMatureVsReprofileTest
  */
 
