--- old/test/compiler/jvmci/compilerToVM/LookupMethodInPoolTest.java	2017-04-21 12:15:12.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/LookupMethodInPoolTest.java	2017-04-21 12:15:12.000000000 +0200
@@ -40,6 +40,7 @@
  *                                sun.hotspot.WhiteBox$WhiteBoxPermission
  * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions
  *                   -XX:+WhiteBoxAPI -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI
+ *                   -Djvmci.Compiler=null
  *                   compiler.jvmci.compilerToVM.LookupMethodInPoolTest
  */
 
