--- old/test/compiler/jvmci/common/JVMCIHelpers.java	2016-05-18 23:02:21.000000000 +0200
+++ new/test/compiler/jvmci/common/JVMCIHelpers.java	2016-05-18 23:02:21.000000000 +0200
@@ -41,12 +41,18 @@
         // just empty, using default interface methods
     }
 
+    public static class EmptyCompilationRequestResult implements CompilationRequestResult {
+        @Override
+        public Object getFailure() {
+            return "no compiler configured";
+        }
+    }
     public static class EmptyHotspotCompiler implements JVMCICompiler {
 
         @Override
         public CompilationRequestResult compileMethod(CompilationRequest request) {
             // do nothing
-            return CompilationRequestResult.failure("no compiler configured", true);
+            return new EmptyCompilationRequestResult();
         }
     }
 
