--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedJavaField.java	2015-12-26 16:48:26.000000000 +0100
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedJavaField.java	2015-12-26 16:48:26.000000000 +0100
@@ -40,17 +40,14 @@
     int offset();
 
     /**
-     * Checks if this field has the {@link Stable} annotation.
-     *
-     * @return true if field has {@link Stable} annotation, false otherwise
+     * Determines if this field should be treated as a constant.
      */
     boolean isStable();
 
     /**
-     * If this field is stable, checks if default values (0, null, etc.) should be considered stable
-     * as well.
-     *
-     * @return true if default values should be considered stable, false otherwise
+     * Determines if this field should be considered constant if it has the default value for its
+     * type (e.g, 0, null, etc.). This result of this method is undefined if this field is not
+     * {@linkplain #isStable() stable}.
      */
     boolean isDefaultStable();
 }
