--- old/src/share/vm/opto/memnode.cpp	2014-01-20 23:43:39.593523778 +0100
+++ new/src/share/vm/opto/memnode.cpp	2014-01-20 23:43:39.529523778 +0100
@@ -1763,6 +1763,7 @@
             }
             // Add a dependence for invalidation of the optimization.
             if (!call_site->is_constant_call_site()) {
+              VM_ENTRY_MARK;
               C->dependencies()->assert_call_site_target_value(call_site, target);
             }
             return t;
