--- old/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedJavaFieldImpl.java	2018-01-29 17:58:44.000000000 +0100
+++ new/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotResolvedJavaFieldImpl.java	2018-01-29 17:58:44.000000000 +0100
@@ -28,7 +28,6 @@
 import java.lang.annotation.Annotation;
 import java.lang.reflect.Field;
 
-import jdk.internal.vm.annotation.Stable;
 import jdk.vm.ci.meta.JavaType;
 import jdk.vm.ci.meta.ResolvedJavaType;
 
@@ -143,9 +142,10 @@
     }
 
     /**
-     * Checks if this field has the {@link Stable} annotation.
+     * Checks if this field has the {@code jdk.internal.vm.annotation.Stable} annotation.
      *
-     * @return true if field has {@link Stable} annotation, false otherwise
+     * @return true if field has {@code jdk.internal.vm.annotation.Stable} annotation, false
+     *         otherwise
      */
     public boolean isStable() {
         return (config().jvmAccFieldStable & modifiers) != 0;
