--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.meta/src/jdk/vm/ci/meta/Local.java	2016-06-22 17:25:48.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.meta/src/jdk/vm/ci/meta/Local.java	2016-06-22 17:25:48.000000000 +0200
@@ -22,6 +22,9 @@
  */
 package jdk.vm.ci.meta;
 
+/**
+ * Describes the type and bytecode index range in which a local variable is live.
+ */
 public class Local {
 
     private final String name;
