--- old/src/hotspot/share/runtime/jniHandles.cpp	2017-10-30 11:59:36.000000000 +0100
+++ new/src/hotspot/share/runtime/jniHandles.cpp	2017-10-30 11:59:36.000000000 +0100
@@ -141,6 +141,11 @@
 }
 
 
+bool JNIHandles::is_global_weak_cleared(jweak handle) {
+  assert(is_jweak(handle), "not a weak handle");
+  return guard_value<false>(jweak_ref(handle)) == NULL;
+}
+
 void JNIHandles::oops_do(OopClosure* f) {
   f->do_oop(&_deleted_handle);
   _global_handles->oops_do(f);
