--- old/test/jdk/jdk/modules/etc/UpgradeableModules.java	2018-04-18 12:57:32.000000000 +0200
+++ new/test/jdk/jdk/modules/etc/UpgradeableModules.java	2018-04-18 12:57:32.000000000 +0200
@@ -46,6 +46,7 @@
         List.of("java.compiler",
                 "java.jnlp",
                 "jdk.internal.vm.compiler",
+                "jdk.internal.vm.compiler.management",
                 "jdk.deploy",
                 "jdk.javaws",
                 "jdk.plugin",
--- /dev/null	2018-04-18 12:57:33.000000000 +0200
+++ new/src/jdk.internal.vm.compiler.management/share/classes/org.graalvm.compiler.hotspot.management/src/org/graalvm/compiler/hotspot/management/HotSpotGraalManagement.java	2018-04-18 12:57:33.000000000 +0200
@@ -0,0 +1,29 @@
+/*
+ * Copyright (c) 2018, 2018, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.graalvm.compiler.hotspot.management;
+
+/**
+ * Placeholder until next Graal update.
+ */
+public final class HotSpotGraalManagement  {
+}
