--- old/src/share/vm/jvmci/jvmciRuntime.hpp	2015-12-22 15:26:59.000000000 +0100
+++ new/src/share/vm/jvmci/jvmciRuntime.hpp	2015-12-22 15:26:59.000000000 +0100
@@ -71,8 +71,6 @@
   static bool _HotSpotJVMCIRuntime_initialized;
   static bool _well_known_classes_initialized;
   static const char* _compiler;
-  static int _options_count;
-  static SystemProperty** _options;
 
   static int _trivial_prefixes_count;
   static char** _trivial_prefixes;
@@ -99,20 +97,6 @@
    */
   static void save_compiler(const char* compiler);
 
-  /**
-   * Saves the value of the system properties starting with "jvmci.option." for processing
-   * when JVMCI is initialized.
-   *
-   * @param props the head of the system property list
-   */
-  static void save_options(SystemProperty* props);
-
-  /**
-   * If either the PrintFlags or ShowFlags JVMCI option is present,
-   * then JVMCI is initialized to show the help message.
-   */
-  static void maybe_print_flags(TRAPS);
-
   static bool is_HotSpotJVMCIRuntime_initialized() { return _HotSpotJVMCIRuntime_initialized; }
 
   /**
