--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot.aarch64/src/jdk/vm/ci/hotspot/aarch64/AArch64HotSpotVMConfig.java	2016-06-22 17:25:37.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot.aarch64/src/jdk/vm/ci/hotspot/aarch64/AArch64HotSpotVMConfig.java	2016-06-22 17:25:37.000000000 +0200
@@ -36,11 +36,6 @@
         super(config);
     }
 
-    /**
-     * Maximum allowed size of allocated area for a frame.
-     */
-    final int maxFrameSize = 16 * 1024;
-
     final boolean linuxOs = System.getProperty("os.name", "").startsWith("Linux");
 
     final boolean useCompressedOops = getFlag("UseCompressedOops", Boolean.class);
