--- old/src/share/vm/jvmci/jvmci_globals.cpp	2016-08-11 15:25:49.000000000 +0200
+++ new/src/share/vm/jvmci/jvmci_globals.cpp	2016-08-11 15:25:49.000000000 +0200
@@ -84,6 +84,7 @@
   CHECK_NOT_SET(JVMCICountersExcludeCompiler, EnableJVMCI)
   CHECK_NOT_SET(JVMCIUseFastLocking,          EnableJVMCI)
   CHECK_NOT_SET(JVMCINMethodSizeLimit,        EnableJVMCI)
+  CHECK_NOT_SET(MethodProfileWidth,           EnableJVMCI)
   CHECK_NOT_SET(TraceUncollectedSpeculations, EnableJVMCI)
 
 #ifndef PRODUCT
--- old/src/share/vm/jvmci/jvmci_globals.hpp	2016-08-11 15:25:50.000000000 +0200
+++ new/src/share/vm/jvmci/jvmci_globals.hpp	2016-08-11 15:25:50.000000000 +0200
@@ -88,6 +88,9 @@
   experimental(intx, JVMCINMethodSizeLimit, (80*K)*wordSize,                \
           "Maximum size of a compiled method.")                             \
                                                                             \
+  experimental(intx, MethodProfileWidth, 0,                                 \
+          "Number of methods to record in call profile")                    \
+                                                                            \
   develop(bool, TraceUncollectedSpeculations, false,                        \
           "Print message when a failed speculation was not collected")
 
--- old/src/share/vm/oops/methodData.hpp	2016-08-11 15:25:50.000000000 +0200
+++ new/src/share/vm/oops/methodData.hpp	2016-08-11 15:25:50.000000000 +0200
@@ -30,6 +30,9 @@
 #include "oops/method.hpp"
 #include "oops/oop.hpp"
 #include "runtime/orderAccess.hpp"
+#if INCLUDE_JVMCI
+#include "jvmci/jvmci_globals.hpp"
+#endif
 
 class BytecodeStream;
 class KlassSizeStats;
--- old/src/share/vm/runtime/globals.hpp	2016-08-11 15:25:51.000000000 +0200
+++ new/src/share/vm/runtime/globals.hpp	2016-08-11 15:25:51.000000000 +0200
@@ -3027,9 +3027,6 @@
           "Number of receiver types to record in call/cast profile")        \
           range(0, 8)                                                       \
                                                                             \
-  experimental(intx, MethodProfileWidth, 0,                                 \
-          "Number of methods to record in call profile")                    \
-                                                                            \
   develop(intx, BciProfileWidth,      2,                                    \
           "Number of return bci's to record in ret profile")                \
                                                                             \
