--- old/src/hotspot/share/gc/g1/g1SATBCardTableModRefBS.hpp	2017-11-02 08:30:05.000000000 +0100
+++ new/src/hotspot/share/gc/g1/g1SATBCardTableModRefBS.hpp	2017-11-02 08:30:05.000000000 +0100
@@ -95,9 +95,6 @@
     jbyte val = _byte_map[card_index];
     return (val & (clean_card_mask_val() | deferred_card_val())) == deferred_card_val();
   }
-  virtual void write_ref_nmethod_pre(oop* dst, nmethod* nm);
-  virtual void write_ref_nmethod_post(oop* dst, nmethod* nm);
-
 };
 
 template<>
