--- old/src/share/vm/jvmci/jvmciCompiler.cpp	2016-06-23 12:39:22.000000000 +0200
+++ new/src/share/vm/jvmci/jvmciCompiler.cpp	2016-06-23 12:39:22.000000000 +0200
@@ -125,6 +125,11 @@
 
 void JVMCICompiler::compile_method(const methodHandle& method, int entry_bci, JVMCIEnv* env) {
   JVMCI_EXCEPTION_CONTEXT
+  if (SystemDictionary::java_system_loader() == NULL) {
+    // 8160121: The system class loader is cached in the VM after VM init phase 3 has completed.
+    // We cannot compile before then as the ServiceLoader cannot be used until after phase 3.
+    return;
+  }
 
   bool is_osr = entry_bci != InvocationEntryBci;
   if (_bootstrapping && is_osr) {
