--- old/src/share/vm/runtime/compilationPolicy.hpp	2016-01-11 13:55:48.000000000 +0100
+++ new/src/share/vm/runtime/compilationPolicy.hpp	2016-01-11 13:55:48.000000000 +0100
@@ -58,6 +58,8 @@
   static void set_policy(CompilationPolicy* policy) { _policy = policy; }
   static CompilationPolicy* policy()                { return _policy; }
 
+  static CompileTask* select_task_helper(CompileQueue* compile_queue);
+
   // Profiling
   elapsedTimer* accumulated_time() { return &_accumulated_time; }
   void print_time() PRODUCT_RETURN;
