--- old/src/share/vm/runtime/simpleThresholdPolicy.cpp	2016-01-11 13:55:49.000000000 +0100
+++ new/src/share/vm/runtime/simpleThresholdPolicy.cpp	2016-01-11 13:55:49.000000000 +0100
@@ -168,7 +168,7 @@
 
 // Called with the queue locked and with at least one element
 CompileTask* SimpleThresholdPolicy::select_task(CompileQueue* compile_queue) {
-  return compile_queue->first();
+  return select_task_helper(compile_queue);
 }
 
 void SimpleThresholdPolicy::reprofile(ScopeDesc* trap_scope, bool is_osr) {
