--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotForeignCallTarget.java	2016-06-22 17:25:41.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotForeignCallTarget.java	2016-06-22 17:25:41.000000000 +0200
@@ -29,6 +29,7 @@
     /**
      * The entry point address of this call's target.
      */
+    @SuppressFBWarnings(value = "URF_UNREAD_PUBLIC_OR_PROTECTED_FIELD", justification = "accessed by subclasses")//
     protected long address;
 
     public HotSpotForeignCallTarget(long address) {
