--- old/src/share/vm/runtime/globals.hpp	2014-02-03 11:38:10.000000000 +0100
+++ new/src/share/vm/runtime/globals.hpp	2014-02-03 11:38:09.000000000 +0100
@@ -3840,8 +3840,8 @@
   product(bool , AllowNonVirtualCalls, false,                               \
           "Obey the ACC_SUPER flag and allow invokenonvirtual calls")       \
                                                                             \
-  product(bool, UseHSAILSimulator, false,                                   \
-          "Run code on HSAIL Simulator")                                    \
+  product(bool, UseHSAILSimulator, true,                                    \
+          "Use HSAIL simulator if available (fall back to HSAIL hardware)") \
                                                                             \
   diagnostic(ccstr, SharedArchiveFile, NULL,                                \
           "Override the default location of the CDS archive file")          \
