--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.meta/src/jdk/vm/ci/meta/ResolvedJavaField.java	2016-09-05 18:31:18.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.meta/src/jdk/vm/ci/meta/ResolvedJavaField.java	2016-09-05 18:31:18.000000000 +0200
@@ -52,7 +52,9 @@
     /**
      * Determines if this field is a synthetic field as defined by the Java Language Specification.
      */
-    boolean isSynthetic();
+    default boolean isSynthetic() {
+        return (SYNTHETIC & getModifiers()) == SYNTHETIC;
+    }
 
     /**
      * Returns the {@link ResolvedJavaType} object representing the class or interface that declares
