--- old/src/hotspot/share/jvmci/jvmciJavaClasses.hpp	2018-11-01 11:13:10.000000000 +0100
+++ new/src/hotspot/share/jvmci/jvmciJavaClasses.hpp	2018-11-01 11:13:10.000000000 +0100
@@ -278,6 +278,7 @@
   end_class                                                                                                                                                    \
   start_class(VirtualObject)                                                                                                                                   \
     int_field(VirtualObject, id)                                                                                                                               \
+    oop_field(VirtualObject, baseObject, "Ljdk/vm/ci/meta/JavaValue;")                                                                                         \
     oop_field(VirtualObject, type, "Ljdk/vm/ci/meta/ResolvedJavaType;")                                                                                        \
     objArrayOop_field(VirtualObject, values, "[Ljdk/vm/ci/meta/JavaValue;")                                                                                    \
     objArrayOop_field(VirtualObject, slotKinds, "[Ljdk/vm/ci/meta/JavaKind;")                                                                                  \
