--- old/src/java.management/share/classes/module-info.java	2018-04-18 11:08:32.000000000 +0200
+++ new/src/java.management/share/classes/module-info.java	2018-04-18 11:08:32.000000000 +0200
@@ -64,8 +64,7 @@
     exports sun.management.counter.perf to
         jdk.management.agent;
     exports sun.management.spi to
-        jdk.management,
-        jdk.internal.vm.compiler.management;
+        jdk.management;
 
     uses javax.management.remote.JMXConnectorProvider;
     uses javax.management.remote.JMXConnectorServerProvider;
