--- old/src/share/vm/compiler/compileBroker.cpp	2016-07-13 17:32:41.000000000 +0200
+++ new/src/share/vm/compiler/compileBroker.cpp	2016-07-13 17:32:41.000000000 +0200
@@ -1037,6 +1037,12 @@
   AbstractCompiler *comp = CompileBroker::compiler(comp_level);
   assert(comp != NULL, "Ensure we have a compiler");
 
+#if INCLUDE_JVMCI
+  if (comp->is_jvmci() && !JVMCIRuntime::can_initialize_JVMCI()) {
+    return NULL;
+  }
+#endif
+
   DirectiveSet* directive = DirectivesStack::getMatchingDirective(method, comp);
   nmethod* nm = CompileBroker::compile_method(method, osr_bci, comp_level, hot_method, hot_count, compile_reason, directive, THREAD);
   DirectivesStack::release(directive);
