--- old/src/hotspot/share/jvmci/jvmciCompilerToVM.cpp	2018-08-29 16:58:48.000000000 +0200
+++ new/src/hotspot/share/jvmci/jvmciCompilerToVM.cpp	2018-08-29 16:58:48.000000000 +0200
@@ -1059,11 +1059,10 @@
               } else {
                 // some object might already have been re-allocated, only reallocate the non-allocated ones
                 objects = new GrowableArray<ScopeValue*>(scope->objects()->length());
-                int ii = 0;
                 for (int i = 0; i < scope->objects()->length(); i++) {
                   ObjectValue* sv = (ObjectValue*) scope->objects()->at(i);
                   if (sv->value().is_null()) {
-                    objects->at_put(ii++, sv);
+                    objects->append(sv);
                   }
                 }
               }
