--- old/src/share/vm/jvmci/jvmciCompiler.hpp	2016-05-10 14:10:36.000000000 +0200
+++ new/src/share/vm/jvmci/jvmciCompiler.hpp	2016-05-10 14:10:36.000000000 +0200
@@ -68,8 +68,14 @@
   // Initialization
   virtual void initialize();
 
+  /**
+   * Initialize the compile queue with the methods in java.lang.Object and
+   * then wait until the queue is empty.
+   */
   void bootstrap();
 
+  bool is_bootstrapping() const { return _bootstrapping; }
+
   // Compilation entry point for methods
   virtual void compile_method(ciEnv* env, ciMethod* target, int entry_bci, DirectiveSet* directive);
 
