--- old/test/compiler/jvmci/events/JvmciShutdownEventTest.java	2016-06-23 12:39:25.000000000 +0200
+++ new/test/compiler/jvmci/events/JvmciShutdownEventTest.java	2016-06-23 12:39:25.000000000 +0200
@@ -40,6 +40,8 @@
  * @run main ClassFileInstaller
  *     compiler.jvmci.common.JVMCIHelpers$EmptyHotspotCompiler
  *     compiler.jvmci.common.JVMCIHelpers$EmptyCompilerFactory
+ *     compiler.jvmci.common.JVMCIHelpers$EmptyCompilationRequestResult
+ *     compiler.jvmci.common.JVMCIHelpers$EmptyVMEventListener
  *     compiler.jvmci.events.JvmciShutdownEventListener
  * @run main/othervm compiler.jvmci.events.JvmciShutdownEventTest
  */
