--- old/test/compiler/jvmci/events/JvmciNotifyBootstrapFinishedEventTest.java	2016-06-23 12:39:24.000000000 +0200
+++ new/test/compiler/jvmci/events/JvmciNotifyBootstrapFinishedEventTest.java	2016-06-23 12:39:24.000000000 +0200
@@ -43,6 +43,8 @@
  * @run main ClassFileInstaller
  *     compiler.jvmci.common.JVMCIHelpers$EmptyHotspotCompiler
  *     compiler.jvmci.common.JVMCIHelpers$EmptyCompilerFactory
+ *     compiler.jvmci.common.JVMCIHelpers$EmptyCompilationRequestResult
+ *     compiler.jvmci.common.JVMCIHelpers$EmptyVMEventListener
  *     compiler.jvmci.events.JvmciNotifyBootstrapFinishedEventTest
  *     jdk.test.lib.Asserts
  *     jdk.test.lib.Utils
