--- old/src/share/vm/jvmci/jvmciEnv.cpp	2016-08-12 18:23:34.000000000 +0200
+++ new/src/share/vm/jvmci/jvmciEnv.cpp	2016-08-12 18:23:34.000000000 +0200
@@ -342,7 +342,7 @@
   Symbol* sig_sym  = cpool->signature_ref_at(index);
 
   if (cpool->has_preresolution()
-      || (holder() == SystemDictionary::MethodHandle_klass() &&
+      || ((holder() == SystemDictionary::MethodHandle_klass() || holder() == SystemDictionary::VarHandle_klass()) &&
           MethodHandles::is_signature_polymorphic_name(holder(), name_sym))) {
     // Short-circuit lookups for JSR 292-related call sites.
     // That is, do not rely only on name-based lookups, because they may fail
