--- old/src/hotspot/share/code/debugInfo.cpp	2018-11-01 11:13:06.000000000 +0100
+++ new/src/hotspot/share/code/debugInfo.cpp	2018-11-01 11:13:05.000000000 +0100
@@ -127,6 +127,7 @@
 
 void ObjectValue::read_object(DebugInfoReadStream* stream) {
   _klass = read_from(stream);
+  _base_object = read_from(stream);
   assert(_klass->is_constant_oop(), "should be constant java mirror oop");
   int length = stream->read_int();
   for (int i = 0; i < length; i++) {
@@ -144,6 +145,7 @@
     stream->write_int(OBJECT_CODE);
     stream->write_int(_id);
     _klass->write_on(stream);
+    _base_object->write_on(stream);
     int length = _field_values.length();
     stream->write_int(length);
     for (int i = 0; i < length; i++) {
