--- old/src/hotspot/share/code/debugInfo.hpp	2018-11-01 11:13:07.000000000 +0100
+++ new/src/hotspot/share/code/debugInfo.hpp	2018-11-01 11:13:06.000000000 +0100
@@ -97,14 +97,16 @@
  private:
   int                        _id;
   ScopeValue*                _klass;
+  ScopeValue*                _base_object;
   GrowableArray<ScopeValue*> _field_values;
   Handle                     _value;
   bool                       _visited;
 
  public:
-  ObjectValue(int id, ScopeValue* klass)
+  ObjectValue(int id, ScopeValue* klass, ScopeValue* base_object)
      : _id(id)
      , _klass(klass)
+     , _base_object(base_object)
      , _field_values()
      , _value()
      , _visited(false) {
@@ -122,6 +124,7 @@
   bool                        is_object() const         { return true; }
   int                         id() const                { return _id; }
   ScopeValue*                 klass() const             { return _klass; }
+  ScopeValue*                 base_object() const       { return _base_object; }
   GrowableArray<ScopeValue*>* field_values()            { return &_field_values; }
   ScopeValue*                 field_at(int i) const     { return _field_values.at(i); }
   int                         field_size()              { return _field_values.length(); }
