--- old/src/share/vm/runtime/arguments.cpp	2017-04-28 11:49:05.000000000 +0200
+++ new/src/share/vm/runtime/arguments.cpp	2017-04-28 11:49:05.000000000 +0200
@@ -2515,6 +2515,9 @@
   status = status && check_jvmci_args_consistency();
 
   if (EnableJVMCI) {
+    PropertyList_unique_add(&_system_properties, "jdk.internal.vm.ci.enabled", "true",
+        AddProperty, UnwriteableProperty, InternalProperty);
+
     if (!ScavengeRootsInCode) {
       warning("forcing ScavengeRootsInCode non-zero because JVMCI is enabled");
       ScavengeRootsInCode = 1;
