--- old/src/share/vm/jvmci/vmStructs_jvmci.cpp	2016-06-29 14:06:25.000000000 +0200
+++ new/src/share/vm/jvmci/vmStructs_jvmci.cpp	2016-06-29 14:06:25.000000000 +0200
@@ -550,7 +550,7 @@
   declare_function(os::javaTimeNanos)                                     \
                                                                           \
   declare_function(Deoptimization::fetch_unroll_info)                     \
-  COMPILER2_PRESENT(declare_function(Deoptimization::uncommon_trap))      \
+  declare_function(Deoptimization::uncommon_trap)                         \
   declare_function(Deoptimization::unpack_frames)                         \
                                                                           \
   declare_function(JVMCIRuntime::new_instance) \
