--- old/src/hotspot/share/jvmci/jvmciRuntime.cpp	2017-11-03 18:40:52.000000000 +0100
+++ new/src/hotspot/share/jvmci/jvmciRuntime.cpp	2017-11-03 18:40:52.000000000 +0100
@@ -821,7 +821,7 @@
 }
 
 CompLevel JVMCIRuntime::adjust_comp_level_inner(const methodHandle& method, bool is_osr, CompLevel level, JavaThread* thread) {
-  JVMCICompiler* compiler = JVMCICompiler::instance(thread);
+  JVMCICompiler* compiler = JVMCICompiler::instance(false, thread);
   if (compiler != NULL && compiler->is_bootstrapping()) {
     return level;
   }
