--- old/test/compiler/jvmci/errors/CodeInstallerTest.java	2016-06-22 17:25:49.000000000 +0200
+++ new/test/compiler/jvmci/errors/CodeInstallerTest.java	2016-06-22 17:25:49.000000000 +0200
@@ -28,6 +28,7 @@
 import jdk.vm.ci.code.Architecture;
 import jdk.vm.ci.code.CodeCacheProvider;
 import jdk.vm.ci.code.Register;
+import jdk.vm.ci.code.RegisterArray;
 import jdk.vm.ci.code.StackSlot;
 import jdk.vm.ci.code.site.DataPatch;
 import jdk.vm.ci.code.site.Site;
@@ -80,11 +81,11 @@
 
     protected Register getRegister(PlatformKind kind, int index) {
         int idx = index;
-        Register[] allRegs = arch.getAvailableValueRegisters();
-        for (int i = 0; i < allRegs.length; i++) {
-            if (arch.canStoreValue(allRegs[i].getRegisterCategory(), kind)) {
+        RegisterArray allRegs = arch.getAvailableValueRegisters();
+        for (Register reg : allRegs) {
+            if (arch.canStoreValue(reg.getRegisterCategory(), kind)) {
                 if (idx-- == 0) {
-                    return allRegs[i];
+                    return reg;
                 }
             }
         }
