--- old/src/hotspot/share/runtime/sharedRuntime.hpp	2020-05-07 10:59:44.000000000 +0200
+++ new/src/hotspot/share/runtime/sharedRuntime.hpp	2020-05-07 10:59:44.000000000 +0200
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1997, 2018, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2020, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -334,6 +334,11 @@
   // Find the method that called us.
   static methodHandle find_callee_method(JavaThread* thread, TRAPS);
 
+  static void monitor_enter_helper(oopDesc* obj, BasicLock* lock, JavaThread* thread,
+                                   bool use_inlined_fast_locking);
+
+  static void monitor_exit_helper(oopDesc* obj, BasicLock* lock, JavaThread* thread,
+                                  bool use_inlined_fast_locking);
 
  private:
   static Handle find_callee_info(JavaThread* thread,
