--- old/test/compiler/jvmci/compilerToVM/GetVtableIndexForInterfaceTest.java	2017-08-23 23:10:13.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/GetVtableIndexForInterfaceTest.java	2017-08-23 23:10:13.000000000 +0200
@@ -131,7 +131,7 @@
         System.out.println(tcase);
         Method method = tcase.holder.getDeclaredMethod(tcase.methodName);
         HotSpotResolvedObjectType metaspaceKlass = CompilerToVMHelper
-                .lookupType(Utils.toJVMTypeSignature(tcase.receiver),
+                .lookupTypeHelper(Utils.toJVMTypeSignature(tcase.receiver),
                         getClass(), /* resolve = */ true);
         HotSpotResolvedJavaMethod metaspaceMethod = CTVMUtilities
                 .getResolvedMethod(tcase.holder, method);
