--- old/src/share/vm/jvmci/jvmciCodeInstaller.cpp	2016-05-18 23:02:17.000000000 +0200
+++ new/src/share/vm/jvmci/jvmciCodeInstaller.cpp	2016-05-18 23:02:17.000000000 +0200
@@ -495,7 +495,7 @@
     return result;
   }
 
-  _debug_recorder->pcs_size(); // ehm, create the sentinel record
+  _debug_recorder->pcs_size(); // create the sentinel record
 
   assert(_debug_recorder->pcs_length() >= 2, "must be at least 2");
 
