--- old/test/java/lang/Class/getDeclaredField/FieldSetAccessibleTest.java	2017-04-28 20:31:31.000000000 +0200
+++ new/test/java/lang/Class/getDeclaredField/FieldSetAccessibleTest.java	2017-04-28 20:31:31.000000000 +0200
@@ -286,7 +286,11 @@
          * Filter deployment modules
          */
         static Set<String> systemModules() {
-            Set<String> mods = Set.of("javafx.deploy", "jdk.deploy", "jdk.plugin", "jdk.javaws");
+            Set<String> mods = Set.of("javafx.deploy", "jdk.deploy", "jdk.plugin", "jdk.javaws",
+                // All JVMCI packages other than jdk.vm.ci.services are dynamically
+                // exported to jdk.internal.vm.compiler and jdk.aot
+                "jdk.internal.vm.compiler", "jdk.aot"
+            );
             return ModuleFinder.ofSystem().findAll().stream()
                                .map(mref -> mref.descriptor().name())
                                .filter(mn -> !mods.contains(mn))
