--- old/.mx.jvmci/suite.py	2016-01-11 15:00:53.000000000 +0100
+++ new/.mx.jvmci/suite.py	2016-01-11 15:00:53.000000000 +0100
@@ -301,32 +301,5 @@
         "JVMCI_SERVICE",
       ],
     },
-
-    # This exists to have a monolithic jvmci.jar file which simplifies
-    # using the -Xoverride option in JDK9.
-    "JVMCI" : {
-      "subDir" : "src/jdk.vm.ci/share/classes",
-      "overlaps" : [
-        "JVMCI_API",
-        "JVMCI_SERVICE",
-        "JVMCI_HOTSPOT",
-        "JVMCI_HOTSPOTVMCONFIG",
-        "JVMCI_SERVICE_PROCESSOR",
-      ],
-      "dependencies" : [
-        "jdk.vm.ci.service",
-        "jdk.vm.ci.inittimer",
-        "jdk.vm.ci.runtime",
-        "jdk.vm.ci.common",
-        "jdk.vm.ci.aarch64",
-        "jdk.vm.ci.amd64",
-        "jdk.vm.ci.sparc",
-        "jdk.vm.ci.hotspotvmconfig",
-        "jdk.vm.ci.hotspot.aarch64",
-        "jdk.vm.ci.hotspot.amd64",
-        "jdk.vm.ci.hotspot.sparc",
-        "jdk.vm.ci.service.processor"
-      ],
-    },
   },
 }
