--- old/src/share/vm/jvmci/vmStructs_jvmci.cpp	2016-08-23 18:57:32.000000000 +0200
+++ new/src/share/vm/jvmci/vmStructs_jvmci.cpp	2016-08-23 18:57:32.000000000 +0200
@@ -493,6 +493,7 @@
   declare_constant(Method::_force_inline)                                 \
   declare_constant(Method::_dont_inline)                                  \
   declare_constant(Method::_hidden)                                       \
+  declare_constant(Method::_intrinsic_candidate)                          \
   declare_constant(Method::_reserved_stack_access)                        \
                                                                           \
   declare_constant(Method::nonvirtual_vtable_index)                       \
