--- old/src/share/vm/compiler/compileTask.cpp	2016-01-11 13:55:46.000000000 +0100
+++ new/src/share/vm/compiler/compileTask.cpp	2016-01-11 13:55:46.000000000 +0100
@@ -91,6 +91,7 @@
   _osr_bci = osr_bci;
   _is_blocking = is_blocking;
   JVMCI_ONLY(_has_waiter = CompileBroker::compiler(comp_level)->is_jvmci();)
+  JVMCI_ONLY(_jvmci_compiler_thread = NULL;)
   _comp_level = comp_level;
   _num_inlined_bytecodes = 0;
 
