--- old/src/share/vm/runtime/arguments.cpp	2015-12-22 15:26:59.000000000 +0100
+++ new/src/share/vm/runtime/arguments.cpp	2015-12-22 15:26:59.000000000 +0100
@@ -3344,12 +3344,6 @@
   const char* fileSep = os::file_separator();
   sprintf(path, "%s%slib%sendorsed", Arguments::get_java_home(), fileSep, fileSep);
 
-#if INCLUDE_JVMCI
-  if (EnableJVMCI) {
-    JVMCIRuntime::save_options(_system_properties);
-  }
-#endif // INCLUDE_JVMCI
-
   if (CheckEndorsedAndExtDirs) {
     int nonEmptyDirs = 0;
     // check endorsed directory
