--- old/test/compiler/jvmci/compilerToVM/GetClassInitializerTest.java	2017-04-20 21:50:10.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/GetClassInitializerTest.java	2017-04-20 21:50:10.000000000 +0200
@@ -31,6 +31,7 @@
  * @modules jdk.internal.vm.ci/jdk.vm.ci.hotspot
  * @build jdk.internal.vm.ci/jdk.vm.ci.hotspot.CompilerToVMHelper
  * @run main/othervm -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI
+ *                   -Djvmci.Compiler=null
  *                   compiler.jvmci.compilerToVM.GetClassInitializerTest
  */
 
