--- old/test/compiler/jvmci/compilerToVM/GetExceptionTableTest.java	2017-04-20 21:50:12.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/GetExceptionTableTest.java	2017-04-20 21:50:11.000000000 +0200
@@ -34,6 +34,7 @@
  *          jdk.internal.vm.ci/jdk.vm.ci.code
  * @build jdk.internal.vm.ci/jdk.vm.ci.hotspot.CompilerToVMHelper
  * @run main/othervm -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI
+ *                   -Djvmci.Compiler=null
  *                   compiler.jvmci.compilerToVM.GetExceptionTableTest
  */
 
@@ -137,4 +138,3 @@
         }
     }
 }
-
