--- old/test/compiler/jvmci/compilerToVM/LookupKlassInPoolTest.java	2017-04-20 21:50:23.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/LookupKlassInPoolTest.java	2017-04-20 21:50:22.000000000 +0200
@@ -42,6 +42,7 @@
  * @run main/othervm -Xbootclasspath/a:.
  *                   -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI
  *                   -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI
+ *                   -Djvmci.Compiler=null
  *                   compiler.jvmci.compilerToVM.LookupKlassInPoolTest
  */
 
