--- old/test/compiler/jvmci/compilerToVM/ReprofileTest.java	2017-04-20 21:50:30.000000000 +0200
+++ new/test/compiler/jvmci/compilerToVM/ReprofileTest.java	2017-04-20 21:50:29.000000000 +0200
@@ -40,7 +40,7 @@
  * @run main/othervm -Xbootclasspath/a:.
  *                   -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI
  *                   -XX:+UnlockExperimentalVMOptions -XX:+EnableJVMCI
- *                   -Xmixed -Xbatch
+ *                   -Xmixed -Xbatch -Djvmci.Compiler=null
  *                   compiler.jvmci.compilerToVM.ReprofileTest
  */
 
