--- old/src/jdk.vm.ci/share/classes/jdk.vm.ci.meta/src/jdk/vm/ci/meta/LocalVariableTable.java	2016-06-21 23:03:57.000000000 +0200
+++ new/src/jdk.vm.ci/share/classes/jdk.vm.ci.meta/src/jdk/vm/ci/meta/LocalVariableTable.java	2016-06-21 23:03:57.000000000 +0200
@@ -25,10 +25,19 @@
 import java.util.ArrayList;
 import java.util.List;
 
+/**
+ * @see "https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.13"
+ */
 public class LocalVariableTable {
 
     private final Local[] locals;
 
+    /**
+     *
+     * @param locals array of objects describing local variables. This array is now owned by this
+     *            object and must not be mutated by the caller.
+     */
+    @SuppressFBWarnings(value = "EI_EXPOSE_REP2", justification = "caller transfers ownership of `locals`")
     public LocalVariableTable(Local[] locals) {
         this.locals = locals;
     }
@@ -47,10 +56,6 @@
         return result;
     }
 
-    public Local[] getLocals() {
-        return locals;
-    }
-
     public Local[] getLocalsAt(int bci) {
         List<Local> result = new ArrayList<>();
         for (Local l : locals) {
