--- old/src/share/vm/c1/c1_GraphBuilder.cpp	2014-01-20 23:43:38.341523789 +0100
+++ new/src/share/vm/c1/c1_GraphBuilder.cpp	2014-01-20 23:43:38.285523788 +0100
@@ -1621,6 +1621,7 @@
                   constant = new Constant(as_ValueType(field_val));
                   // Add a dependence for invalidation of the optimization.
                   if (!call_site->is_constant_call_site()) {
+                    VM_ENTRY_MARK;
                     dependency_recorder()->assert_call_site_target_value(call_site, target);
                   }
                 }
