--- old/src/cpu/x86/vm/templateInterpreterGenerator_x86.cpp	2016-09-05 18:47:23.000000000 +0200
+++ new/src/cpu/x86/vm/templateInterpreterGenerator_x86.cpp	2016-09-05 18:47:23.000000000 +0200
@@ -55,7 +55,7 @@
 // Run with +PrintInterpreter to get the VM to print out the size.
 // Max size with JVMTI
 #ifdef AMD64
-int TemplateInterpreter::InterpreterCodeSize = 256 * 1024;
+int TemplateInterpreter::InterpreterCodeSize = JVMCI_ONLY(264) NOT_JVMCI(256) * 1024;
 #else
 int TemplateInterpreter::InterpreterCodeSize = 224 * 1024;
 #endif // AMD64
