--- old/src/share/vm/runtime/arguments.cpp	2016-07-15 01:10:45.000000000 +0200
+++ new/src/share/vm/runtime/arguments.cpp	2016-07-15 01:10:45.000000000 +0200
@@ -2276,11 +2276,7 @@
 #if INCLUDE_JVMCI
 // Check consistency of jvmci vm argument settings.
 bool Arguments::check_jvmci_args_consistency() {
-  if (!EnableJVMCI && !JVMCIGlobals::check_jvmci_flags_are_consistent()) {
-    JVMCIGlobals::print_jvmci_args_inconsistency_error_message();
-    return false;
-  }
-  return true;
+   return JVMCIGlobals::check_jvmci_flags_are_consistent();
 }
 #endif //INCLUDE_JVMCI
 
