--- old/src/os/windows/vm/gpu_windows.cpp	2014-02-03 11:38:09.000000000 +0100
+++ new/src/os/windows/vm/gpu_windows.cpp	2014-02-03 11:38:09.000000000 +0100
@@ -28,7 +28,7 @@
 
 jobject gpu::probe_gpus(JNIEnv* env) {
   // TODO: add detection of PTX/NVidia
-  if (UseHSAILSimulator && gpu::Hsail::register_natives(env)) {
+  if (gpu::Hsail::register_natives(env)) {
     return env->NewStringUTF("HSAIL");
   }
   return env->NewStringUTF("");
