--- old/src/share/vm/compiler/compileTask.cpp	2015-11-10 08:18:00.000000000 +0100
+++ new/src/share/vm/compiler/compileTask.cpp	2015-11-10 08:18:00.000000000 +0100
@@ -90,6 +90,7 @@
   _method_holder = JNIHandles::make_global(method->method_holder()->klass_holder());
   _osr_bci = osr_bci;
   _is_blocking = is_blocking;
+  JVMCI_ONLY(_has_waiter = CompileBroker::compiler(comp_level)->is_jvmci();)
   _comp_level = comp_level;
   _num_inlined_bytecodes = 0;
 
