--- old/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotConstantPoolObject.java	2018-08-15 23:16:28.000000000 +0200
+++ new/src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotConstantPoolObject.java	2018-08-15 23:16:28.000000000 +0200
@@ -23,11 +23,10 @@
 package jdk.vm.ci.hotspot;
 
 import jdk.vm.ci.meta.JavaConstant;
-import jdk.vm.ci.meta.ResolvedJavaType;
 
 /**
- * Represents a constant that was retrieved from a constant pool.
- * Used to keep track of the constant pool slot for the constant.
+ * Represents a constant that was retrieved from a constant pool. Used to keep track of the constant
+ * pool slot for the constant.
  */
 public final class HotSpotConstantPoolObject extends HotSpotObjectConstantImpl {
 
@@ -36,14 +35,19 @@
     }
 
     public static JavaConstant forObject(HotSpotResolvedObjectType type, int cpi, JavaConstant object) {
-        return forObject(type, cpi, ((HotSpotObjectConstantImpl)object).object());
+        return forObject(type, cpi, ((HotSpotObjectConstantImpl) object).object());
     }
 
     private final HotSpotResolvedObjectType type;
     private final int cpi;
 
-    public HotSpotResolvedObjectType getCpType() { return type; }
-    public int getCpi()  { return cpi; }
+    public HotSpotResolvedObjectType getCpType() {
+        return type;
+    }
+
+    public int getCpi() {
+        return cpi;
+    }
 
     HotSpotConstantPoolObject(HotSpotResolvedObjectType type, int cpi, Object object) {
         super(object, false);
